model/proxy/alphav1/config/cfg.proto

```diff
@@ -21,230 +21,162 @@ package istio.proxy.v1alpha.config;
-  // Faults include aborting the Http request from downstream service,
-  // delaying the proxying of request to the upstream service, or both.
-  // TODO: This should move into UpstreamCluster
-  HttpFaultInjection fault = 3;
-
-  // Custom properties per rule (depends on proxy), such as open tracing,
-  // access log formats, etc. Some of these will be defined in future
-  // iterations.
-  google.protobuf.Any custom_impl = 4;
+  // L4 fault injection policy applies to L4 traffic
+  L4FaultInjection l4_fault = 7;
+
+  // Custom policy implementations
+  google.protobuf.Any custom = 8;
 }
 
-// Basic routing rule match criterion using Tcp attributes and downstream
-// cluster identifier
-message L4MatchCondition {
-  // Set of layer 4 match attributes such as src ip/port, dst ip/port and
-  // protocol
-  L4MatchAttributes l4attributes = 1;
+// Route rule provides a custom routing policy based on the source and
+// destination service versions and connection/request metadata.  The rule must
+// provide a set of conditions for each protocol (TCP, UDP, HTTP) that the
+// destination service exposes on its ports. The rule applies only to the ports
+// on the destination service for which it provides protocol-specific match
+// condition, e.g. if the rule does not specify TCP condition, the rule does
+// not apply to TCP traffic towards the destination service.
+message RouteRule {
+  // Uniquely identifies the destination service (not its versions) associated
+  // with this routing rule. This is applicable for hostname-based resolution for
+  // HTTP traffic as well as IP-based resolution for TCP/UDP traffic.  *required*
+  string destination = 1;
+
+  // Identifies the service initiating a connection or a request.
+  string source = 2;
+
+  // Identifies the source service version. The identifier is interpreted
+  // by the platform to match a service version for the source service.
+  //
+  // N.B. The map is used instead of pstruct due to lack of serialization support
+  // in golang protobuf library (see https://github.com/golang/protobuf/pull/208)
+  map<string, string> source_version = 3;
 
-  // Identify the downstream cluster initiating the connection.
-  ClusterIdentifier src_cluster = 2;
+  // Set of layer 4 match conditions based on the IP ranges
+  L4MatchAttributes tcp = 4;
+
+  // Set of layer 4 match conditions based on the IP ranges
+  L4MatchAttributes udp = 5;
+
+  // Set of HTTP match conditions based on the request metadata
+  HTTPMatchAttributes http = 6;
+
+  // Each routing rule is associated with one or more service versions (see
+  // glossary in beginning of document). Weights associated with the service
+  // version determine the proportion of traffic it receives.
+  repeated DestinationWeight versions = 7;
+
+  // Precedence is used to disambiguate the order of application of rules
+  // for the same destination service. A higher number takes priority.
+  int32 precedence = 8;
```

Comment:
This is nice. The rules format is evolving into the existing amalgam8 constructs @frankb. We can't escape precedence. 

I would also add a version field to the rule. So that it gives the end user the ability to rollback to a stable rule if they mess up something. this would certainly mean that manager needs to provide a rollback API. (Even if you don't want to do this now, can we just add this comment so that we can track this issue. Matter of fact, a CDN company I spoke to actually wanted this capability).

I would add twl more lines here.
 E.g. for http, routes under a virtual host need to be ordered. Here precedence is priority. 

Secondly, depending on the platform, the storage for the rules format may not support atomic replace (and concurrently will be a bigger mess). The easiest way to tackle this is to ask the user to create a new rule with higher priority and delete the old rule with lower priority. Priority essentially here becomes a rule version of some sort. E.g. send all traffic to helloworld/hello to v1. Then next rule is split traffic to helloworld/hello between v1 and v2. Both rules apply to same service and same route and virtual host. 

To provide atomic update semantics to the user, the platform must support a transactional storage and patch/edit semantics as well. K8s may have etcd but unsure if mesos has.